Mathlib/Analysis/NormedSpace/Star/Matrix.lean

```diff
@@ -70,3 +88,122 @@ theorem entrywise_sup_norm_bound_of_unitary {U : Matrix n n 𝕜} (hU : U ∈ Ma
+  @NormedRing.induced ((Matrix n n 𝕜) ≃⋆ₐ[𝕜] (EuclideanSpace 𝕜 n →L[𝕜] EuclideanSpace 𝕜 n))
+    _ _ _ ContinuousLinearMap.toNormedRing _ _ toEuclideanClm.injective
+
+open Bornology Filter
+open scoped Topology Uniformity
+
+/-- The metric on `Matrix m n 𝕜` arising from the operator norm given by the identification with
+(continuous) linear maps of `EuclideanSpace`. -/
+def instMetricSpaceL2Op : MetricSpace (Matrix m n 𝕜) := by
+  /- We first replace the topology so that we can automatically replace the uniformity using
+  `UniformAddGroup.toUniformSpace_eq`. -/
+  letI normed_add_comm_group : NormedAddCommGroup (Matrix m n 𝕜) :=
+    { normedAddCommGroupAuxL2Op.replaceTopology <|
+        (toEuclideanLin (𝕜 := 𝕜) (m := m) (n := n)).trans toContinuousLinearMap
+        |>.toContinuousLinearEquiv.toHomeomorph.inducing.induced with
+      norm := normedAddCommGroupAuxL2Op.norm
+      dist_eq := normedAddCommGroupAuxL2Op.dist_eq }
+  exact normed_add_comm_group.replaceUniformity <| by
+    congr
+    rw [← @UniformAddGroup.toUniformSpace_eq _ (instUniformSpaceMatrix m n 𝕜) _ _]
+    rw [@UniformAddGroup.toUniformSpace_eq _ PseudoEMetricSpace.toUniformSpace _ _]
+
+scoped[Matrix.L2OpNorm] attribute [instance] Matrix.instMetricSpaceL2Op
+
+open scoped Matrix.L2OpNorm
+
+/-- The norm structure on `Matrix m n 𝕜` arising from the operator norm given by the identification
+with (continuous) linear maps of `EuclideanSpace`. -/
+def instNormedAddCommGroupL2Op : NormedAddCommGroup (Matrix m n 𝕜) where
+  norm := normedAddCommGroupAuxL2Op.norm
+  dist_eq := normedAddCommGroupAuxL2Op.dist_eq
+
+scoped[Matrix.L2OpNorm] attribute [instance] Matrix.instNormedAddCommGroupL2Op
+
+lemma op_norm_def (x : Matrix m n 𝕜) :
+    ‖x‖ = ‖(toEuclideanLin (𝕜 := 𝕜) (m := m) (n := n)).trans toContinuousLinearMap x‖ := rfl
+
+-- note: with only a type ascription in the left-hand side, Lean picks the wrong norm.
+lemma norm_mulVec_le (A : Matrix m n 𝕜) (x : EuclideanSpace 𝕜 n) :
+    ‖(EuclideanSpace.equiv m 𝕜).symm <| A.mulVec x‖ ≤ ‖A‖ * ‖x‖ :=
+  toEuclideanLin (n := n) (m := m) (𝕜 := 𝕜) |>.trans toContinuousLinearMap A |>.le_op_norm x
+
+/-- The normed algebra structure on `Matrix n n 𝕜` arising from the operator norm given by the
+identification with (continuous) linear endmorphisms of `EuclideanSpace 𝕜 n`. -/
+def instNormedSpaceL2Op : NormedSpace 𝕜 (Matrix m n 𝕜) where
+  norm_smul_le r x := by
+    rw [op_norm_def, LinearEquiv.map_smul]
+    exact norm_smul_le r ((toEuclideanLin (𝕜 := 𝕜) (m := m) (n := n)).trans toContinuousLinearMap x)
+
+scoped[Matrix.L2OpNorm] attribute [instance] Matrix.instNormedSpaceL2Op
+
+/-- The normed ring structure on `Matrix n n 𝕜` arising from the operator norm given by the
+identification with (continuous) linear endmorphisms of `EuclideanSpace 𝕜 n`. -/
+def instNormedRingL2Op : NormedRing (Matrix n n 𝕜) where
+  dist_eq := normedRingAuxL2Op.dist_eq
+  norm_mul := normedRingAuxL2Op.norm_mul
```

Comment:
Can you prove this result for rectangular matrices like we do for the other matrix norms?